modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,385 @@
+                "Friend already added for this challenge",
+            )));
+        }
+    }
+
+    existing_friends.extend(friends);
+    CHALLENGE_FRIENDS.save(deps.storage, challenge_id, &existing_friends)?;
+    Ok(Response::new().add_attribute("action", "add_friends"))
+}
+
+fn daily_check_in(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: String,
+    metadata: Option<String>,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    let today = date_from_block(&env);
+
+    // Check if Admin has already checked in today
+    if DAILY_CHECK_INS
+        .load(deps.storage, today.clone())
+        .map_or(false, |check_in| {
+            check_in.last_checked_in.as_deref() == Some(&today)
+        })
+    {
+        return Err(AppError::AlreadyCheckedIn {});
+    }
+
+    let _blocks_per_day = 1440; // dummy value, check this
+    let next_check_in_block = env.block.height + 10;
+
+    let check_in = CheckIn {
+        last_checked_in: Some(today.clone()),
+        next_check_in_by: next_check_in_block,
+        metadata,
+    };
+
+    DAILY_CHECK_INS.save(deps.storage, challenge_id, &check_in)?;
+    Ok(Response::new().add_attribute("action", "check_in"))
+}
+
+fn cast_vote(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    _app: &ChallengeApp,
+    vote: Option<bool>,
+    challenge_id: &str,
+) -> AppResult {
+    let today = date_from_block(&env);
+
+    // If Admin checked in today, friends can't vote
+    if DAILY_CHECK_INS
+        .load(deps.storage, today.clone())
+        .map_or(false, |check_in| {
+            check_in.last_checked_in.as_deref() == Some(&today)
+        })
+    {
+        return Err(AppError::AlreadyCheckedIn {});
+    }
+
+    // If the vote is None, default to true (meaning we assume Admin fulfilled his challenge)
+    let final_vote = vote.unwrap_or(true);
+
+    // Construct the vote
+    let vote_entry = Vote {
+        voter: info.sender.to_string(),
+        vote: Some(final_vote),
+        challenge_id: challenge_id.to_owned(),
+    };
+
+    // Load existing votes for the current block height or initialize an empty list if none exist
+    let mut votes_for_block = VOTES
+        .load(deps.storage, challenge_id.to_owned())
+        .unwrap_or_else(|_| Vec::new());
+
+    // Append the new vote
+    votes_for_block.push(vote_entry);
+
+    // Save the updated votes
+    VOTES.save(deps.storage, challenge_id.to_owned(), &votes_for_block)?;
+
+    Ok(Response::new().add_attribute("action", "cast_vote"))
+}
+
+fn count_votes(
+    deps: DepsMut,
+    info: MessageInfo,
+    _env: Env,
+    app: &ChallengeApp,
+    challenge_id: String,
+) -> AppResult {
+    let votes_for_challenge = VOTES
+        .load(deps.storage, challenge_id.clone())
+        .unwrap_or_else(|_| Vec::new());
+
+    let any_false_vote = votes_for_challenge.iter().any(|v| v.vote == Some(false));
```

Comment:
This could get abused by his friends, maybe some treshold would be appropriate? And/or the ability for the admin to veto a count. In the end it's the admin's good will so if he really deems his opinion thruthworthy he should be able to overwrite his friend's descision. The ability for him to do so itself will push his friend to collaborate.